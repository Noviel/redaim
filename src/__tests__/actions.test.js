import { 
  createAction,
  createActionR,
  
  wrapActions,
  
  createActions,
  createActionsR,
  
  bindActions,
  bindActionsR

} from '../actions';

describe('createAction', () => {  
  
  it('should return function when called once', () => {
    expect(typeof createAction()).toMatchSnapshot();
  });

  it('should return function when called twice', () => {
    expect(typeof createAction()()).toMatchSnapshot();
  });  
  
  it('should return object when called thrice', () => {
    expect(typeof createAction()()()).toMatchSnapshot();
  });

  it('should set trigger to default value if not specified', () => {
    const { type } = createAction()()();
    
    expect(type).toMatchSnapshot();
  });

  it('should set target to default value if not specified', () => {
    const { meta: { target } } = createAction()()();
    
    expect(target).toMatchSnapshot();
  });

  it('should set exec to default value if not specified', () => {
    const { meta: { exec } } = createAction()()();
    
    expect(exec).toMatchSnapshot();
  });

  it('should set action\'s type', () => {
    const { type } = createAction('ACTION_TRIGGER')()();

    expect(type).toMatchSnapshot();
  });

  it('should set action\'s meta.target', () => {
    const { meta: { target } } = createAction('ACTION_TRIGGER')('target')();

    expect(target).toMatchSnapshot();
  });

  it('should correctly merge action\'s properties', () => {
    const action = createAction('ACTION_TRIGGER')('target')({ 
      exec: 'EXEC_TYPE',
      payload: { data: 10 },
      meta: {
        obj: {
          inner: 10
        },
        value: 100
      }
    });

    expect(action).toMatchSnapshot();
  });

  it('should override action\'s type', () => {
    const action = createAction('TYPE_FROM_TRIGGER')('target')({
      type: 'TYPE_OVERRIDE'
    });

    expect(action.type).toMatchSnapshot();
  });

  it('should override action\'s target', () => {
    const action = createAction('TYPE_FROM_TRIGGER')('target')({
      meta: { target: 'target_override' }
    });

    expect(action.meta.target).toMatchSnapshot();
  });

  it('should correctly merge custom action\'s properties', () => {
    const actionOne = createAction()()({
      customProp: {
        value: 42
      }
    });

    const actionTwo = createAction()()({
      payload: {
        data: 'payload-data'
      },
      meta: {
        val: 'meta-val'
      },
      customProp: {
        value: 42
      }
    });
    
    const toMatch = {
      actionOne,
      actionTwo
    };

    expect(toMatch).toMatchSnapshot();
  });

});

describe('createActionR', () => {  
  
  it('should return function when called once', () => {
    expect(typeof createActionR()).toMatchSnapshot();
  });

  it('should return function when called twice', () => {
    expect(typeof createActionR()()).toMatchSnapshot();
  });  
  
  it('should return object when called thrice', () => {
    expect(typeof createActionR()()()).toMatchSnapshot();
  });

  it('should set trigger to default value if not specified', () => {
    const { type } = createActionR()()();
    
    expect(type).toMatchSnapshot();
  });

  it('should set target to default value if not specified', () => {
    const { meta: { target } } = createActionR()()();
    
    expect(target).toMatchSnapshot();
  });

  it('should set exec to default value if not specified', () => {
    const { meta: { exec } } = createActionR()()();
    
    expect(exec).toMatchSnapshot();
  });

  it('should set action\'s type', () => {
    const { type } = createActionR()('ACTION_TRIGGER')();

    expect(type).toMatchSnapshot();
  });

  it('should set action\'s meta.target', () => {
    const { meta: { target } } = createActionR()('ACTION_TRIGGER')('target');

    expect(target).toMatchSnapshot();
  });

  it('should correctly merge action\'s properties', () => {
    const action = createActionR({ 
      exec: 'EXEC_TYPE',
      payload: { data: 10 },
      meta: {
        obj: {
          inner: 10
        },
        value: 100
      }
    })('ACTION_TRIGGER')('target');

    expect(action).toMatchSnapshot();
  });

  it('should override action\'s type', () => {
    const action = createActionR({
      type: 'TYPE_OVERRIDE'
    })('TYPE_FROM_TRIGGER')('target');

    expect(action.type).toMatchSnapshot();
  });

  it('should override action\'s target', () => {
    const action = createActionR({
      meta: { target: 'target_override' }
    })('TYPE_FROM_TRIGGER')('target');

    expect(action.meta.target).toMatchSnapshot();
  });

  it('should correctly merge custom action\'s properties', () => {
    const actionOne = createActionR({
      customProp: {
        value: 42
      }
    })()();

    const actionTwo = createActionR({
      payload: {
        data: 'payload-data'
      },
      meta: {
        val: 'meta-val'
      },
      customProp: {
        value: 42
      }
    })()();
    
    const toMatch = {
      actionOne,
      actionTwo
    };

    expect(toMatch).toMatchSnapshot();
  });

});

describe('wrapActions', () => {
  const actions = {
    'ACTION_ONE': {
      type: 'ACTION_ONE'
    },
    'ACTION_TWO': {
      type: 'ACTION_TWO'
    }
  };

  const identity = x => x;

  it('should return object with same keys as an input action', () => {
    const result = wrapActions(identity)(actions);
    const keys = Object.keys(result);

    expect(keys).toMatchSnapshot();
  });

  it('should return object with functions', () => {
    const result = wrapActions(identity)(actions);
    const keys = Object.keys(result);
    
    const toMatch = keys.reduce((acc, curr) => { 
      acc[curr] = typeof result[curr]; 
      return acc; 
    }, {});

    expect(toMatch).toMatchSnapshot();
  });

  it('should correctly apply function without arguments to each action', () => {
    const addConstant = action => ({ ...action, payload: 42 });
    const wrappedActions = wrapActions(addConstant)(actions);

    const wrappedActionOne = wrappedActions['ACTION_ONE']();
    const wrappedActionTwo = wrappedActions['ACTION_TWO']();

    expect({ wrappedActionOne, wrappedActionTwo }).toMatchSnapshot();
  });

  it('should correctly apply function with one argument to each action', () => {
    const addPayload = (action, payload) => ({ ...action, payload });
    const wrappedActions = wrapActions(addPayload)(actions);

    const wrappedActionOne = wrappedActions['ACTION_ONE'](10);
    const wrappedActionTwo = wrappedActions['ACTION_TWO']({ value: { inner: { innerValue: 100 }}});

    expect({ wrappedActionOne, wrappedActionTwo }).toMatchSnapshot();
  });

  it('should correctly apply function with multiple arguments to each action', () => {
    const addOneTwoToPayload = (action, one, two) => ({ ...action, payload: { one, two } });
    const wrappedActions = wrapActions(addOneTwoToPayload)(actions);

    const wrappedActionOne = wrappedActions['ACTION_ONE'](1, 'two');
    const wrappedActionTwo = wrappedActions['ACTION_TWO']({ value: { inner: { innerValue: 100 }}}, null);

    expect({ wrappedActionOne, wrappedActionTwo }).toMatchSnapshot();
  });

});

describe('createActions', () => {
  it('should create actions from object with object as values', () => {
    const actionsObject = {
      exec1: { exec: 'EXEC1', payload: { value: 1 } },
      exec2: { exec: 'EXEC2', payload: { value: 2 } },
      exec3: { exec: 'EXEC3', payload: { value: 3 } }
    };

    const { exec1, exec2, exec3 } = createActions('TRIGGER')('target')(actionsObject);

    const result = {
      exec1: exec1(),
      exec2: exec2(),
      exec3: exec3()
    };

    expect(result).toMatchSnapshot();
  });

  it('should create actions from object with functions as values', () => {
    const actionsObject = {
      exec1: value => ({ exec: 'EXEC1', payload: { value } }),
      exec2: (value1, value2) => ({ exec: 'EXEC2', payload: { value1, value2 } }),
    };

    const { exec1, exec2 } = createActions('TRIGGER')('target')(actionsObject);

    const result = {
      exec1: exec1(100),
      exec2: exec2({ value: 'value'}, 420),
    };

    expect(result).toMatchSnapshot();
  });

  it('should create actions from object with mixed values', () => {
    const actionsObject = {
      exec1: value => ({ exec: 'EXEC1', payload: { value } }),
      exec2: { exec: 'EXEC2', payload: 'payload' },
    };

    const { exec1, exec2 } = createActions('TRIGGER')('target')(actionsObject);

    const result = {
      exec1: exec1('exec1-value'),
      exec2: exec2(),
    };

    expect(result).toMatchSnapshot();
  });
});

describe('createActionsR', () => {
  it('should create actions from object with object as values', () => {
    const actionsObject = {
      exec1: { exec: 'EXEC1', payload: { value: 1 } },
      exec2: { exec: 'EXEC2', payload: { value: 2 } },
      exec3: { exec: 'EXEC3', payload: { value: 3 } }
    };

    const { exec1, exec2, exec3 } = createActionsR(actionsObject)('TRIGGER')('target');

    const result = {
      exec1: exec1(),
      exec2: exec2(),
      exec3: exec3()
    };

    expect(result).toMatchSnapshot();
  });

  it('should create actions from object with functions as values', () => {
    const actionsObject = {
      exec1: value => ({ exec: 'EXEC1', payload: { value } }),
      exec2: (value1, value2) => ({ exec: 'EXEC2', payload: { value1, value2 } }),
    };

    const { exec1, exec2 } = createActionsR(actionsObject)('TRIGGER')('target');

    const result = {
      exec1: exec1(100),
      exec2: exec2({ value: 'value'}, 420),
    };

    expect(result).toMatchSnapshot();
  });

  it('should create actions from object with mixed values', () => {
    const actionsObject = {
      exec1: value => ({ exec: 'EXEC1', payload: { value } }),
      exec2: { exec: 'EXEC2', payload: 'payload' },
    };

    const { exec1, exec2 } = createActionsR(actionsObject)('TRIGGER')('target');

    const result = {
      exec1: exec1('exec1-value'),
      exec2: exec2(),
    };

    expect(result).toMatchSnapshot();
  });
});

describe('bindActions', () => {
  it('should bind trigger to actions', () => {
    const actionsObject = {
      exec1: { exec: 'EXEC1', payload: { value: 1 } },
      exec2: { exec: 'EXEC2', payload: { value: 2 } },
      exec3: { exec: 'EXEC3', payload: { value: 3 } }
    };

    const { exec1, exec2, exec3 } = bindActions(createAction('TRIGGER'))('special_target')(actionsObject);

    const result = {
      exec1: exec1().type,
      exec2: exec2().type,
      exec3: exec3().type
    };

    expect(result).toMatchSnapshot();
  });

  it('should bind target to actions', () => {
    const actionsObject = {
      exec1: { exec: 'EXEC1', payload: { value: 1 } },
      exec2: { exec: 'EXEC2', payload: { value: 2 } },
      exec3: { exec: 'EXEC3', payload: { value: 3 } }
    };

    const { exec1, exec2, exec3 } = bindActions(createAction('TRIGGER'))('special_target')(actionsObject);

    const result = {
      exec1: exec1().meta.target,
      exec2: exec2().meta.target,
      exec3: exec3().meta.target
    };

    expect(result).toMatchSnapshot();
  });
});

describe('bindActionsR', () => {
  it('should bind trigger to actions', () => {
    const actionsObject = {
      exec1: { exec: 'EXEC1', payload: { value: 1 } },
      exec2: { exec: 'EXEC2', payload: { value: 2 } },
      exec3: { exec: 'EXEC3', payload: { value: 3 } }
    };
    const { exec1, exec2, exec3 } = bindActionsR(actionsObject)('TRIGGER')('special_target');

    const result = {
      exec1: exec1().type,
      exec2: exec2().type,
      exec3: exec3().type,
    };

    expect(result).toMatchSnapshot();
  });

  it('should bind target to actions', () => {
    const actionsObject = {
      exec1: { exec: 'EXEC1', payload: { value: 1 } },
      exec2: { exec: 'EXEC2', payload: { value: 2 } },
      exec3: { exec: 'EXEC3', payload: { value: 3 } }
    };

    const { exec1, exec2, exec3 } = bindActionsR(actionsObject)('TRIGGER')('special_target');

    const result = {
      exec1: exec1().meta.target,
      exec2: exec2().meta.target,
      exec3: exec3().meta.target
    };

    expect(result).toMatchSnapshot();
  });
});
