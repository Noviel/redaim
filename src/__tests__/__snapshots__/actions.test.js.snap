// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bindActions should bind target to actions 1`] = `
Object {
  "exec1": "special_target",
  "exec2": "special_target",
  "exec3": "special_target",
}
`;

exports[`bindActions should bind trigger to actions 1`] = `
Object {
  "exec1": "TRIGGER",
  "exec2": "TRIGGER",
  "exec3": "TRIGGER",
}
`;

exports[`bindActionsR should bind target to actions 1`] = `
Object {
  "exec1": "special_target",
  "exec2": "special_target",
  "exec3": "special_target",
}
`;

exports[`bindActionsR should bind trigger to actions 1`] = `
Object {
  "exec1": "TRIGGER",
  "exec2": "TRIGGER",
  "exec3": "TRIGGER",
}
`;

exports[`createAction should correctly merge action's properties 1`] = `
Object {
  "meta": Object {
    "exec": "EXEC_TYPE",
    "obj": Object {
      "inner": 10,
    },
    "target": "target",
    "value": 100,
  },
  "payload": Object {
    "data": 10,
  },
  "type": "ACTION_TRIGGER",
}
`;

exports[`createAction should correctly merge custom action's properties 1`] = `
Object {
  "actionOne": Object {
    "customProp": Object {
      "value": 42,
    },
    "meta": Object {
      "exec": "@@REDAIM-ACTION/UNKNOWN",
      "target": "@@REDAIM-TARGET/UNKNOWN",
    },
    "payload": null,
    "type": "@@REDAIM-TRIGGER/UNKNOWN",
  },
  "actionTwo": Object {
    "customProp": Object {
      "value": 42,
    },
    "meta": Object {
      "exec": "@@REDAIM-ACTION/UNKNOWN",
      "target": "@@REDAIM-TARGET/UNKNOWN",
      "val": "meta-val",
    },
    "payload": Object {
      "data": "payload-data",
    },
    "type": "@@REDAIM-TRIGGER/UNKNOWN",
  },
}
`;

exports[`createAction should override action's target 1`] = `"target_override"`;

exports[`createAction should override action's type 1`] = `"TYPE_OVERRIDE"`;

exports[`createAction should return function when called once 1`] = `"function"`;

exports[`createAction should return function when called twice 1`] = `"function"`;

exports[`createAction should return object when called thrice 1`] = `"object"`;

exports[`createAction should set action's meta.target 1`] = `"target"`;

exports[`createAction should set action's type 1`] = `"ACTION_TRIGGER"`;

exports[`createAction should set exec to default value if not specified 1`] = `"@@REDAIM-ACTION/UNKNOWN"`;

exports[`createAction should set target to default value if not specified 1`] = `"@@REDAIM-TARGET/UNKNOWN"`;

exports[`createAction should set trigger to default value if not specified 1`] = `"@@REDAIM-TRIGGER/UNKNOWN"`;

exports[`createActionR should correctly merge action's properties 1`] = `
Object {
  "meta": Object {
    "exec": "EXEC_TYPE",
    "obj": Object {
      "inner": 10,
    },
    "target": "target",
    "value": 100,
  },
  "payload": Object {
    "data": 10,
  },
  "type": "ACTION_TRIGGER",
}
`;

exports[`createActionR should correctly merge custom action's properties 1`] = `
Object {
  "actionOne": Object {
    "customProp": Object {
      "value": 42,
    },
    "meta": Object {
      "exec": "@@REDAIM-ACTION/UNKNOWN",
      "target": "@@REDAIM-TARGET/UNKNOWN",
    },
    "payload": null,
    "type": "@@REDAIM-TRIGGER/UNKNOWN",
  },
  "actionTwo": Object {
    "customProp": Object {
      "value": 42,
    },
    "meta": Object {
      "exec": "@@REDAIM-ACTION/UNKNOWN",
      "target": "@@REDAIM-TARGET/UNKNOWN",
      "val": "meta-val",
    },
    "payload": Object {
      "data": "payload-data",
    },
    "type": "@@REDAIM-TRIGGER/UNKNOWN",
  },
}
`;

exports[`createActionR should override action's target 1`] = `"target_override"`;

exports[`createActionR should override action's type 1`] = `"TYPE_OVERRIDE"`;

exports[`createActionR should return function when called once 1`] = `"function"`;

exports[`createActionR should return function when called twice 1`] = `"function"`;

exports[`createActionR should return object when called thrice 1`] = `"object"`;

exports[`createActionR should set action's meta.target 1`] = `"target"`;

exports[`createActionR should set action's type 1`] = `"ACTION_TRIGGER"`;

exports[`createActionR should set exec to default value if not specified 1`] = `"@@REDAIM-ACTION/UNKNOWN"`;

exports[`createActionR should set target to default value if not specified 1`] = `"@@REDAIM-TARGET/UNKNOWN"`;

exports[`createActionR should set trigger to default value if not specified 1`] = `"@@REDAIM-TRIGGER/UNKNOWN"`;

exports[`createActions should create actions from object with functions as values 1`] = `
Object {
  "exec1": Object {
    "meta": Object {
      "exec": "EXEC1",
      "target": "target",
    },
    "payload": Object {
      "value": 100,
    },
    "type": "TRIGGER",
  },
  "exec2": Object {
    "meta": Object {
      "exec": "EXEC2",
      "target": "target",
    },
    "payload": Object {
      "value1": Object {
        "value": "value",
      },
      "value2": 420,
    },
    "type": "TRIGGER",
  },
}
`;

exports[`createActions should create actions from object with mixed values 1`] = `
Object {
  "exec1": Object {
    "meta": Object {
      "exec": "EXEC1",
      "target": "target",
    },
    "payload": Object {
      "value": "exec1-value",
    },
    "type": "TRIGGER",
  },
  "exec2": Object {
    "meta": Object {
      "exec": "EXEC2",
      "target": "target",
    },
    "payload": "payload",
    "type": "TRIGGER",
  },
}
`;

exports[`createActions should create actions from object with object as values 1`] = `
Object {
  "exec1": Object {
    "meta": Object {
      "exec": "EXEC1",
      "target": "target",
    },
    "payload": Object {
      "value": 1,
    },
    "type": "TRIGGER",
  },
  "exec2": Object {
    "meta": Object {
      "exec": "EXEC2",
      "target": "target",
    },
    "payload": Object {
      "value": 2,
    },
    "type": "TRIGGER",
  },
  "exec3": Object {
    "meta": Object {
      "exec": "EXEC3",
      "target": "target",
    },
    "payload": Object {
      "value": 3,
    },
    "type": "TRIGGER",
  },
}
`;

exports[`createActionsR should create actions from object with functions as values 1`] = `
Object {
  "exec1": Object {
    "meta": Object {
      "exec": "EXEC1",
      "target": "target",
    },
    "payload": Object {
      "value": 100,
    },
    "type": "TRIGGER",
  },
  "exec2": Object {
    "meta": Object {
      "exec": "EXEC2",
      "target": "target",
    },
    "payload": Object {
      "value1": Object {
        "value": "value",
      },
      "value2": 420,
    },
    "type": "TRIGGER",
  },
}
`;

exports[`createActionsR should create actions from object with mixed values 1`] = `
Object {
  "exec1": Object {
    "meta": Object {
      "exec": "EXEC1",
      "target": "target",
    },
    "payload": Object {
      "value": "exec1-value",
    },
    "type": "TRIGGER",
  },
  "exec2": Object {
    "meta": Object {
      "exec": "EXEC2",
      "target": "target",
    },
    "payload": "payload",
    "type": "TRIGGER",
  },
}
`;

exports[`createActionsR should create actions from object with object as values 1`] = `
Object {
  "exec1": Object {
    "meta": Object {
      "exec": "EXEC1",
      "target": "target",
    },
    "payload": Object {
      "value": 1,
    },
    "type": "TRIGGER",
  },
  "exec2": Object {
    "meta": Object {
      "exec": "EXEC2",
      "target": "target",
    },
    "payload": Object {
      "value": 2,
    },
    "type": "TRIGGER",
  },
  "exec3": Object {
    "meta": Object {
      "exec": "EXEC3",
      "target": "target",
    },
    "payload": Object {
      "value": 3,
    },
    "type": "TRIGGER",
  },
}
`;

exports[`wrapActions should correctly apply function with multiple arguments to each action 1`] = `
Object {
  "wrappedActionOne": Object {
    "payload": Object {
      "one": 1,
      "two": "two",
    },
    "type": "ACTION_ONE",
  },
  "wrappedActionTwo": Object {
    "payload": Object {
      "one": Object {
        "value": Object {
          "inner": Object {
            "innerValue": 100,
          },
        },
      },
      "two": null,
    },
    "type": "ACTION_TWO",
  },
}
`;

exports[`wrapActions should correctly apply function with one argument to each action 1`] = `
Object {
  "wrappedActionOne": Object {
    "payload": 10,
    "type": "ACTION_ONE",
  },
  "wrappedActionTwo": Object {
    "payload": Object {
      "value": Object {
        "inner": Object {
          "innerValue": 100,
        },
      },
    },
    "type": "ACTION_TWO",
  },
}
`;

exports[`wrapActions should correctly apply function without arguments to each action 1`] = `
Object {
  "wrappedActionOne": Object {
    "payload": 42,
    "type": "ACTION_ONE",
  },
  "wrappedActionTwo": Object {
    "payload": 42,
    "type": "ACTION_TWO",
  },
}
`;

exports[`wrapActions should return object with functions 1`] = `
Object {
  "ACTION_ONE": "function",
  "ACTION_TWO": "function",
}
`;

exports[`wrapActions should return object with same keys as an input action 1`] = `
Array [
  "ACTION_ONE",
  "ACTION_TWO",
]
`;
